workflow-runtime/src/main/java/com/squareup/workflow1/WorkflowInterceptor.kt

```diff
@@ -165,3 +268,54 @@ internal fun <P, S, O, R> WorkflowInterceptor.intercept(
     override fun toString(): String = "InterceptedWorkflow($workflow, $this@intercept)"
   }
 }
+
+@OptIn(ExperimentalWorkflowApi::class)
+private class InterceptedRenderContext<P, S, O>(
+  private val baseRenderContext: BaseRenderContext<P, S, O>,
+  private val interceptor: RenderContextInterceptor<P, S, O>
+) : BaseRenderContext<P, S, O>, Sink<WorkflowAction<P, S, O>> {
+  override val actionSink: Sink<WorkflowAction<P, S, O>> get() = this
+
+  override fun send(value: WorkflowAction<P, S, O>) {
+    interceptor.onActionSent(value) { interceptedAction ->
+      baseRenderContext.actionSink.send(interceptedAction)
+    }
+  }
+
+  override fun <ChildPropsT, ChildOutputT, ChildRenderingT> renderChild(
+    child: Workflow<ChildPropsT, ChildOutputT, ChildRenderingT>,
+    props: ChildPropsT,
+    key: String,
+    handler: (ChildOutputT) -> WorkflowAction<P, S, O>
+  ): ChildRenderingT =
+    interceptor.onRenderChild(child, props, key, handler) { iChild, iProps, iKey, iHandler ->
+      baseRenderContext.renderChild(iChild, iProps, iKey, iHandler)
+    }
+
+  override fun runningSideEffect(
+    key: String,
+    sideEffect: suspend CoroutineScope.() -> Unit
+  ) {
+    // We don't want to invite the interceptor to shoot itself in the foot
+    // by making mistakes around the `CoroutineScope` receiver, so the sideEffect
+    // method it's given has no receiver. This means it's up to us to provide one, carefully.
+    val withScopeReceiver = suspend {
+      CoroutineScope(activeCoroutineContext()).sideEffect()
+    }
+
+    interceptor.onRunningSideEffect(key, withScopeReceiver) { iKey, iSideEffect ->
+      baseRenderContext.runningSideEffect(iKey) {
+        iSideEffect()
+      }
+    }
+  }
+
+  /**
+   * In a block with a CoroutineScope receiver, calls to `coroutineContext` bind
+   * to `CoroutineScope.coroutineContext` instead of `suspend val coroutineContext`.
+   * Call this and always get the latter.
+   */
+  private suspend fun activeCoroutineContext(): CoroutineContext {
```

Comment:
Make this inline? We just need to force the compiler to call the right method, no need to generate a whole new suspend state machine.